### PR TITLE
Add log_table

### PR DESCRIPTION
**Pull request by irenedea**
# What does this PR do?

TLDR: support the equivalent of log_image for NLP

Researchers have been rolling their own NLP model output loggers that print to console, so seems like we should support this more natively.

# What issue(s) does this change relate to?

[CO-1527](https://mosaicml.atlassian.net/browse/CO-1527)
https://mosaicml.atlassian.net/browse/CO-1527

# Manual Tests

WandBLogger
```
import wandb
wandb.init()
from composer.loggers import WandBLogger

wbl = WandBLogger()
wbl.log_table(columns=["prompt", "generation"], rows=[["I like", "turtles"], ["I love", "MosaicML"]])
```
image

MLFlowLogger
```
from composer.loggers import MLFlowLogger

mfl = MLFlowLogger()
mfl = mfl.log_table(columns=["prompt", "generation"], rows=[["I like", "turtles"], ["I love", "MosaicML"]])

```
image

CometMLLogger
```
from composer.loggers import CometMLLogger
import os
os.environ['COMET_API_KEY'] ="<REDACTED>"

cl = CometMLLogger()
cl.log_table(columns=["prompt", "generation"], rows=[["I like", "turtles"], ["I love", "MosaicML"]])

```
image

FileLogger
```
from composer.loggers import FileLogger
from composer import Trainer
from composer.models import mnist_model


trainer = Trainer(model=mnist_model())

log_destination = FileLogger(
    filename='test.txt')
fl = FileLogger(filename = "test.txt")
fl.init(trainer.state, None)
fl.log_table(columns=["prompt", "generation"], rows=[["I like", "turtles"], ["I love", "MosaicML"]])
fl.file.close()
```
image

ConsoleLogger
```
from composer import Trainer
from composer.models import mnist_model
from composer.loggers import ConsoleLogger
import sys

trainer = Trainer(model=mnist_model())

cl = ConsoleLogger(stream = sys.stdout)
cl.log_table(columns=["prompt", "generation"], rows=[["I like", "turtles"], ["I love", "MosaicML"]])
cl.log_to_console({}, trainer.state)
```

image

InMemoryLogger
```
from composer.loggers import InMemoryLogger

iml = InMemoryLogger()

iml.log_table(columns=["prompt", "generation"], rows=[["I like", "turtles"], ["I love", "MosaicML"]])

print(iml.tables)
```

image

# Before submitting
- [ ] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [ ] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [ ] Did you run the tests locally to make sure they pass?
- [ ] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

[CO-1527]: https://mosaicml.atlassian.net/browse/CO-1527?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by eracah — COMMENTED** [review]
great job! looks really good! thanks for the manual tests
 
Any chance you can add unit tests that just check to make sure the tables are saved? I think we have unit test examples for all 3 loggers that use the run directories to make sure the proper things are saved. See `composer/tests/loggers/*`

**Review by dakinggg — COMMENTED** [review]
Looks good to me!

**Review by dakinggg — COMMENTED** [review]
Should we add this to `ConsoleLogger`, `FileLogger`, and `InMemoryLogger` too?

**Comment by irenedea** [comment]
Added log_table to ConsoleLogger, FileLogger, and InMemoryLogger @dakinggg

File: composer/loggers/console_logger.py [hunk 1]
```diff
@@ -79,6 +80,12 @@ def log_hyperparameters(self, hyperparameters: Dict[str, Any]):
         # Lazy logging of hyperparameters.
         self.hparams.update(hyperparameters)
 
+    def log_table(self, columns: List[str], rows: List[List[Any]], name: str = 'Table') -> None:
+        import pandas as pd
+        table = pd.DataFrame.from_records(data=rows, columns=columns).to_json(orient='split', index=False)
+        self.tables[name] = str(table)
+        del table
```
Comment (dakinggg) [inline composer/loggers/console_logger.py:87]: is this needed? does it not destruct properly?

File: composer/loggers/console_logger.py [hunk 2]
```diff
@@ -203,6 +210,8 @@ def log_to_console(self, data: Dict[str, Any], state: State, prefix: str = '', i
         for data_name, data in data.items():
             data_str = format_log_data_value(data)
             log_str += f'\n\t {prefix}{data_name}: {data_str}'
+        for table_name, table in self.tables.items():
+            log_str += f'\n\t {prefix}{table_name}: {table}'
```
Comment (dakinggg) [inline composer/loggers/console_logger.py:214]: can you make a jira to make this prettier? not priority, just would be nice if it printed an actual table

**Comment by dakinggg** [comment]
I think we'll need to either make pandas a core dependency, or add an optional for table logging or something. @eracah which do you think we should do? Alternatively we could make our "table" representation a dictionary, not sure which is preferred.

File: composer/loggers/console_logger.py (hunk 1 above)
Comment (irenedea) [reply composer/loggers/console_logger.py:87]: good catch! not needed.

File: composer/loggers/console_logger.py (hunk 2 above)
Comment (irenedea) [reply composer/loggers/console_logger.py:214]: https://mosaicml.atlassian.net/browse/GRT-2303

**Comment by eracah** [comment]
> I think we'll need to either make pandas a core dependency, or add an optional for table logging or something. @eracah which do you think we should do? Alternatively we could make our "table" representation a dictionary, not sure which is preferred, but it might be nice to not introduce the pandas dependency.

yeah good q. I think it's fine? llm-foundry requires panda for their eval stuff, which might get upstreamed anyway

**Comment by dakinggg** [comment]
But we still need to add a new optional dep group for pandas right?

**Comment by eracah** [comment]
> But we still need to add a new optional dep group for pandas right?

sure yeah!

File: setup.py [hunk 3]
```diff
@@ -222,6 +222,8 @@ def package_files(prefix: str, directory: str, extension: str):
     'mlflow>=2.0.1,<3.0',
 ]
 
+extra_deps['pandas'] = ['pandas>=2.0.3']
```
Comment (dakinggg) [inline setup.py:225]: do you think we need this minimum pin?

**Review by dakinggg — APPROVED** [review]
LGTM!

File: setup.py (hunk 3 above)
Comment (irenedea) [reply setup.py:225]: Nope! I wasn't too sure about convention, so I started with this, but it seems like it's better to just do equals

Comment (irenedea) [inline setup.py:225]:
```suggestion
extra_deps['pandas'] = ['pandas==2.0.3']
```

Comment (dakinggg) [reply setup.py]: Oh I actually meant we can probably be a lot less restrictive than this since we are only using pandas very minimally. Like maybe just `pandas<3` is ok? or maybe `>=2` and `<3`?

Comment (irenedea) [reply setup.py]: ohhh, gotcha! i'll update it to the latter

File: setup.py [hunk 4]
```diff
@@ -222,6 +222,8 @@ def package_files(prefix: str, directory: str, extension: str):
     'mlflow>=2.0.1,<3.0',
 ]
 
+extra_deps['pandas'] = ['pandas==2.0.3']
```
Comment (irenedea) [inline setup.py:225]:
```suggestion
extra_deps['pandas'] = ['pandas>=2.0.0,<3.0']
```

**Review by eracah — APPROVED** [review]
Great work, @irene-dea! LGTM
